src/coreclr/jit/llvm.cpp

```diff
@@ -841,26 +852,131 @@ llvm::Value* Llvm::buildUserFuncCall(GenTreeCall* call)
-            }
+            llvmFunc = Function::Create(buildHelperLlvmFunctionType(call, true), Function::ExternalLinkage, 0U, symbolName, _module); // TODO: ExternalLinkage forced as defined in ILC module
+        }
 
-            // replacement for _info.compCompHnd->recordRelocation(nullptr, gtCall->gtEntryPoint.handle, IMAGE_REL_BASED_REL32);
-            (*_addCodeReloc)(_thisPtr, call->gtEntryPoint.handle);
+        // replacement for _info.compCompHnd->recordRelocation(nullptr, gtCall->gtEntryPoint.handle, IMAGE_REL_BASED_REL32);
+        (*_addCodeReloc)(_thisPtr, call->gtEntryPoint.handle);
 
-            mapGenTreeToValue(node, _builder.CreateCall(llvmFunc, getShadowStackForCallee()));
-            return;
+        mapGenTreeToValue(call, _builder.CreateCall(llvmFunc, getShadowStackForCallee()));
+        return;
+    }
+    else
+    {
+        fgArgInfo* argInfo = call->fgArgInfo;
+        unsigned int argCount = argInfo->ArgCount();
+        fgArgTabEntry** argTable = argInfo->ArgTable();
+        std::vector<OperandArgNum> sortedArgs = std::vector<OperandArgNum>(argCount);
+        OperandArgNum* sortedData = sortedArgs.data();
+        bool requiresShadowStack = helperRequiresShadowStack(call->gtCallMethHnd);
+
+        for (unsigned i = 0; i < argCount; i++)
+        {
+            fgArgTabEntry* curArgTabEntry = argTable[i];
+            unsigned int   argNum = curArgTabEntry->argNum;
+            OperandArgNum  opAndArg = { argNum, curArgTabEntry->GetNode() };
+            sortedData[argNum] = opAndArg;
+        }
+
+        void* pAddr = nullptr;
+
+        CorInfoHelpFunc helperNum = _compiler->eeGetHelperNum(call->gtCallMethHnd);
+        void* addr = _compiler->compGetHelperFtn(helperNum, &pAddr);
+        const char* symbolName = (*_getMangledSymbolName)(_thisPtr, addr);
+        Function* llvmFunc = _module->getFunction(symbolName);
+        if (llvmFunc == nullptr)
+        {
+            llvmFunc = Function::Create(buildHelperLlvmFunctionType(call, requiresShadowStack), Function::ExternalLinkage, 0U, symbolName, _module);
         }
+
+        (*_addCodeReloc)(_thisPtr, addr);
+
+        std::vector<llvm::Value*> argVec;
+        unsigned argIx = 0;
+
+        Value* shadowStackForCallee = getShadowStackForCallee();
+        if (requiresShadowStack)
+        {
+            argVec.push_back(shadowStackForCallee);
+            argIx++;
+        }
+        else
+        {
+            // we may come back into managed from the unmanaged call so store the shadowstack
+            _builder.CreateStore(shadowStackForCallee, getOrCreateExternalSymbol("t_pShadowStackTop", Type::getInt8PtrTy(_llvmContext)));
+        }
+
+        for (OperandArgNum opAndArg : sortedArgs)
+        {
+            if ((opAndArg.operand->gtOper == GT_CNS_INT) && opAndArg.operand->IsIconHandle())
```

Comment:
```suggestion
            if (opAndArg.operand->IsCnsIntOrI() && opAndArg.operand->IsIconHandle())
```
It seems that we should check for what kind (`GTF_ICON_*`) of handle this is, not assume that it represents a type, though I do not know which flag this one will have.

It also seems that we should register the relocations when building the constant, not when processing the call. It is possible for the constant to get CSEd away, technically.

Reply:
Its not clear what `GTF_ICON_` should be here.  `GTF_ICON_CLASS_HDL` words for some, but for `RhpNewFast` it takes a `MethodTable` and that has `GTF_CALL_VIRT_KIND_MASK` . I've left it without a check for now.

Reply:
`((opAndArg.operand->GetIconHandleFlag() & GTF_ICON_CLASS_HDL) == GTF_ICON_CLASS_HDL)`  "works", but I don't see any indication that `GTF_ICON_*` are bit flags, so that's probably just luck.  I

Reply:
I will dig deeper into this some other day;  let's just leave a `TODO-LLVM` for now.